pyfstat/core.py

```diff
@@ -1007,51 +1149,83 @@ def get_fullycoherent_twoF(
             )
         self.windowRange.t0 = int(tstart)  # TYPE UINT4
         if self.windowRange.tauBand == 0:
-            # true single-template search also in transient params:
-            # actual (t0,tau) window was set with tstart, tend before
             self.windowRange.tau = int(tend - tstart)  # TYPE UINT4
 
         self.FstatMap, self.timingFstatMap = tcw.call_compute_transient_fstat_map(
             self.tCWFstatMapVersion,
             self.tCWFstatMapFeatures,
-            self.FstatResults.multiFatoms[0],
+            self.FstatResults.multiFatoms[0],  # 0 index: single frequency bin
             self.windowRange,
         )
         if self.tCWFstatMapVersion == "lal":
             F_mn = self.FstatMap.F_mn.data
         else:
             F_mn = self.FstatMap.F_mn
 
-        twoF = 2 * np.max(F_mn)
-        if self.BSGL is False:
-            if np.isnan(twoF):
-                return 0
-            else:
-                return twoF
+        # Now instead of the overall twoF,
+        # we get the maximum twoF over the transient window range.
+        self.maxTwoF = 2 * np.max(F_mn)
+        if np.isnan(self.maxTwoF):
+            self.maxTwoF = 0
+        return self.maxTwoF
 
-        FstatResults_single = copy.copy(self.FstatResults)
-        FstatResults_single.numDetectors = 1
-        FstatResults_single.data = self.FstatResults.multiFatoms[0].data[0]
-        FS0 = lalpulsar.ComputeTransientFstatMap(
-            FstatResults_single.multiFatoms[0], self.windowRange, False
-        )
-        FstatResults_single.data = self.FstatResults.multiFatoms[0].data[1]
-        FS1 = lalpulsar.ComputeTransientFstatMap(
-            FstatResults_single.multiFatoms[0], self.windowRange, False
-        )
+    def get_transient_log10BSGL(self):
+        """Computes a transient detection statistic log10BSGL at a single point.
 
-        # for now, use the Doppler parameter with
-        # multi-detector F maximised over t0,tau
-        # to return BSGL
-        # FIXME: should we instead compute BSGL over the whole F_mn
-        # and return the maximum of that?
-        idx_maxTwoF = np.argmax(F_mn)
+        This requires `self.get_transient_maxTwoFstat()` to be run first.
 
-        self.twoFX[0] = 2 * FS0.F_mn.data[idx_maxTwoF]
-        self.twoFX[1] = 2 * FS1.F_mn.data[idx_maxTwoF]
-        log10BSGL = lalpulsar.ComputeBSGL(twoF, self.twoFX, self.BSGLSetup)
+        The single-detector 2F-stat values
+        used for that computation (at the index of `maxTwoF`)
+        are saved in `self.twoFXatMaxTwoF`,
+        not returned.
 
-        return log10BSGL
+        Returns
+        -------
+        log10BSGL: float
+            A single value of the detection statistic log10BSGL
+            at the input parameter values.
+            Also stored as `self.log10BSGL`.
+        """
+        if self.tCWFstatMapVersion == "lal":
+            F_mn = self.FstatMap.F_mn.data
+        else:
+            F_mn = self.FstatMap.F_mn
```

Comment:
suggestion:

```
def __init__(self, yada yada yada, tCWFstatMapVersion="whatever"):
    # ... initi stuff ...
    self.get_tCWFstatMap = {"lal":_get_lal_tCWFstatMap, "pycuda": _get_cuda_tCWFstatMap}

def _get_lal_tCWFstatMap(FstatMap):
    return FstatMap.F_mn.data
    
def _get_cuda_tCWFstatMap(FstatMap):
    return FstatMap.F_mn
```

An now these blocks can be simplified (once for all!) to
```
F_mn = self.get_tCWFstatMap[self.tCWFstatMapVersion](self.FstatMap.F_mn)
```
Instead of a dictionary we can just assign the suitable method and simplify even further.

Reply:
I'm longer-term planning a proper object-based rewrite for the whole "map" stuff anyway.

But I think you're right that just those few lines should work without extra effort.

Reply:
a9d3bb70f538f9a251a413e73df3b0436380eeda